File: flutter/lib/core/models/jobs.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'answers.dart';

part 'jobs.freezed.dart';
part 'jobs.g.dart';

enum JobKind {
  ask,
  codex,
  @JsonValue('kb_reindex')
  kbReindex,
  @JsonValue('paper_ingest')
  paperIngest,
}

enum JobStatus {
  queued,
  running,
  succeeded,
  failed,
  cancelled;

  bool get isActive => this == JobStatus.queued || this == JobStatus.running;
}

@freezed
abstract class JobProgress with _$JobProgress {
  const factory JobProgress({
    @Default('') String stage,
    int? current,
    int? total,
  }) = _JobProgress;

  factory JobProgress.fromJson(Map<String, Object?> json) =>
      _$JobProgressFromJson(json);
}

@freezed
abstract class Job with _$Job {
  const factory Job({
    required String id,
    required JobKind kind,
    required JobStatus status,
    String? userId,
    @Default(<String, dynamic>{}) Map<String, dynamic> params,
    JobProgress? progress,
    JobError? error,
    Map<String, dynamic>? result,
    required DateTime createdAt,
    DateTime? startedAt,
    DateTime? finishedAt,
  }) = _Job;

  factory Job.fromJson(Map<String, Object?> json) => _$JobFromJson(json);
}

File: apple/YAOEKit/Sources/YAOEKit/Models/Jobs.swift
import Foundation

public enum JobKind: String, Codable, Sendable, Hashable {
    case ask
    case codex
    case kbReindex = "kb_reindex"
    case paperIngest = "paper_ingest"
    case answerKb = "answer_kb"
}

public enum JobStatus: String, Codable, Sendable, Hashable {
    case queued
    case running
    case succeeded
    case failed
    case cancelled

    public var isActive: Bool { self == .queued || self == .running }
}

public struct JobProgress: Codable, Sendable, Hashable {
    public var stage: String
    public var current: Int?
    public var total: Int?
}

public struct Job: Codable, Sendable, Hashable, Identifiable {
    public var id: String
    public var kind: JobKind
    public var status: JobStatus
    public var userId: String?
    @DefaultJSONObject public var params: JSONValue
    public var progress: JobProgress?
    public var error: JobError?
    public var result: JSONValue?
    public var createdAt: Date
    public var startedAt: Date?
    public var finishedAt: Date?
}
